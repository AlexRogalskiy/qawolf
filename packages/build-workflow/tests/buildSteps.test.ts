import { loadEvents } from "@qawolf/fixtures";
import { buildClickSteps, buildInputSteps, buildSteps } from "../";

let loginEvents: any[];
let searchAirbnb: any[];

beforeAll(async () => {
  loginEvents = await loadEvents("login");
  searchAirbnb = await loadEvents("airbnb");
});

describe("buildClickSteps", () => {
  it("builds correct steps for login", async () => {
    const steps = buildClickSteps(loginEvents);

    expect(steps.length).toEqual(2);

    // link to form auth
    expect(steps[0].target.xpath).toEqual("//*[@id='content']/ul/li[18]/a");

    // login button
    expect(steps[1].target.xpath).toEqual("//*[@id='login']/button/i");
  });

  it("ignores clicks on date tooltips for search airbnb", async () => {
    const steps = buildClickSteps(searchAirbnb);
    expect(steps.length).toEqual(2);

    // nyc item in dropdown
    expect(steps[0].target.xpath).toEqual(
      "//*[@id='Koan-magic-carpet-koan-search-bar__option-0']/div[2]"
    );

    expect(steps[1].target.inputType).toEqual("submit");
  });
});

describe("buildInputSteps", () => {
  it("builds correct steps for login", async () => {
    const steps = buildInputSteps(loginEvents);
    expect(steps.length).toEqual(2);

    expect(steps[0].target.xpath).toEqual("//*[@id='username']");
    expect(steps[1].target.xpath).toEqual("//*[@id='password']");
  });

  it("builds correct steps for search airbnb", async () => {
    const steps = buildInputSteps(searchAirbnb);

    expect(steps[0].value).toEqual("Fri, Oct 25");
    expect(steps[0].target.xpath).toEqual("//*[@id='checkin_input']");

    expect(steps[1].value).toEqual("Sat, Oct 26");
    expect(steps[1].target.xpath).toEqual("//*[@id='checkout_input']");

    expect(steps[2].value).toEqual("nyc");
    expect(steps[2].target.xpath).toEqual(
      "//*[@id='Koan-magic-carpet-koan-search-bar__input']"
    );
  });
});

describe("buildSteps", () => {
  it("orders events properly", () => {
    const steps = buildSteps(searchAirbnb);

    expect(steps.map(step => step.index)).toEqual([0, 1, 2, 3, 4]);

    expect(steps.map(step => step.action as string)).toEqual([
      "input",
      "input",
      "input",
      "click",
      "click"
    ]);

    expect(steps.map(step => step.target.xpath)).toEqual([
      "//*[@id='checkin_input']",
      "//*[@id='checkout_input']",
      "//*[@id='Koan-magic-carpet-koan-search-bar__input']",
      "//*[@id='Koan-magic-carpet-koan-search-bar__option-0']/div[2]",
      "//*[@id='FMP-target']/div/div[1]/div/div/div/div/div/div[2]/div/form/div[4]/div/button"
    ]);
  });
});
