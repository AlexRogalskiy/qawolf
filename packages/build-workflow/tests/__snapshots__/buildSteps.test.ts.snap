// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildSteps builds expected steps 1`] = `
Array [
  Object {
    "action": "scroll",
    "html": Object {
      "ancestors": Array [],
      "node": Object {
        "attrs": Object {},
        "children": Array [],
        "name": "html",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 0,
    "page": 0,
    "value": Object {
      "x": 0,
      "y": 118,
    },
  },
  Object {
    "action": "click",
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {},
          "children": Array [],
          "name": "li",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {},
          "children": Array [],
          "name": "ul",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "href": "/login",
          "innertext": "Form Authentication",
        },
        "children": Array [
          Object {
            "content": "Form Authentication",
            "type": "text",
          },
        ],
        "name": "a",
        "type": "tag",
        "voidElement": false,
      },
    },
    "index": 1,
    "page": 0,
  },
  Object {
    "action": "type",
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "username",
          "labels": "Username",
          "name": "username",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 2,
    "page": 0,
    "value": "↓KeyT↑KeyT↓KeyO↑KeyO↓KeyM↑KeyM↓KeyS↑KeyS↓KeyM↑KeyM↓KeyI↑KeyI↓KeyT↑KeyT↓KeyH↑KeyH",
  },
  Object {
    "action": "type",
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "username",
          "labels": "Username",
          "name": "username",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 3,
    "page": 0,
    "value": "↓Tab↑Tab",
  },
  Object {
    "action": "type",
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "password",
          "labels": "Password",
          "name": "password",
          "type": "password",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 4,
    "page": 0,
    "value": "SuperSecretPassword!",
  },
  Object {
    "action": "type",
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "password",
          "labels": "Password",
          "name": "password",
          "type": "password",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 5,
    "page": 0,
    "value": "↓Enter↑Enter",
  },
  Object {
    "action": "click",
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "example",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "large-12 columns",
            "id": "content",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "class": "button secondary radius",
          "href": "/logout",
          "innertext": "Logout",
        },
        "children": Array [
          Object {
            "attrs": Object {
              "class": "icon-2x icon-signout",
            },
            "children": Array [
              Object {
                "content": " Logout",
                "type": "text",
              },
            ],
            "name": "i",
            "type": "tag",
            "voidElement": false,
          },
        ],
        "name": "a",
        "type": "tag",
        "voidElement": false,
      },
    },
    "index": 6,
    "page": 0,
  },
]
`;
